src/core/egglog-herbie.rkt

```diff
@@ -854,54 +847,106 @@
 
   (values (reverse all-bindings) curr-bindings))
 
-(define (egglog-unsound-detected curr-program tag params current-schedule)
+(define (egglog-unsound-detected-subprocess tag params egglog-process egglog-output egglog-in err)
+
   (define node-limit (dict-ref params 'node (*node-limit*)))
   (define iter-limit (dict-ref params 'iteration (*default-egglog-iter-limit*)))
 
-  ;; Make a copy here too so that we don't modify our original clean copy
-  (define temp-program (egglog-program-copy curr-program))
-
   ;; Algorithm:
-  ;; 1. Saturate lifting and lowering
+  ;; 1. Run (PUSH) to the save the above state of the egraph
   ;; 2. Repeat rules based on their ruleset tag once
   ;; 3. Run the unsound-rule function ruleset once
   ;; 4. Extract the (unsound) function that returns a bool
-  ;; 5. If (unsound) function returns "true", we have unsoundless -> optimal iter limit is one below this
+  ;; 5. If (unsound) function returns "true", we have unsoundless, so go to Step 10 for ROLLBACK
   ;; 6. Run (print-size) to get nodes of the form "node_name : num_nodes" for all nodes in egraph
-  ;; 7. If the total number of nodes is more than node-limit -> optimal iter limit is one below this
-  ;; 8. Increment until we hit above consition or iter-limit
```

Comment:
Step 7 seems incorrect to me. If we hit the node limit, we stop at that iteration, not rollback by 1.

Reply:
In the egg implementation, when you exceed node limit for the first time, then that is the final iteration. It's worth looking into the performance characteristics of this. Maybe time will double or something ridiculous.